### PR TITLE
MGMT: Add user data to mgmt_handler

### PR DESCRIPTION
**Pull request by kunoh**
PR based on RFC: #60616

Signed-off-by: Kuno Heltborg kunoh@live.dk

### CODE REVIEW

**Review by de-nordic — COMMENTED** [review]
Why not follow with the singleton idea I have proposed in RFC?  It does not require changes to the MCUmgr.

**Comment by kunoh** [comment]
It is possible to solve it using singleton.
But when using it in combination with dependency injection the singleton cannot be created statically and would instead need to be created dynamically. User data solves this.
In general I believe that adding user data would open up to a more freely choice of design patterns.

**Comment by carlescufi** [comment]
@kunoh please rebase
@de-nordic @nordicjm please review

**Comment by de-nordic** [comment]
> @kunoh please rebase @de-nordic @nordicjm please review

It is up to @nordicjm in architect hat to decide on that.
I think that this is not a right approach as current architecture of SMP frame processor works as stream demultiplexer.
There was no need, so far, for such pointer because the user context for methods servicing handlers were working in their own context and knew everything about the environment they are called.
The user case that requires the change is multiplexing the commands back to single point that then again decides how to work with them and needs the extra data stored with the handler.
Note that handler is not something passed between layers of code and needs to carry its context with it, it was assumed that, after demultiplexing, the called method is just passed payload that it parses in its own context.

Assuming that we do not allow anything using this pointer to Zephyr tree then, maybe this is ok, but if we do note that there is one handler per command which means that there is extra sizeof(void *) added with each command, including commands that we already have defined.

Rather than doing such change, I would see a change that would allow handlers to access header fields if needed. This may be more useful, although I do not know how much effort is needed for that, and what would be the memory inpact.

**Comment by kunoh** [comment]
Shall I close this PR?

**Comment by de-nordic** [comment]
> Shall I close this PR?

Wait for architect to respond.

**Comment by kunoh** [comment]
> Not against this change but would wanted it guarded by a Kconfig which is disabled by default if it were to be accepted

I will add a Kconfig

File: subsys/mgmt/mcumgr/mgmt/Kconfig [hunk 1]
```diff
@@ -25,3 +25,10 @@ config MCUMGR_MGMT_NOTIFICATION_HOOKS
 	  or decline the current operation, by returning false this will signal to the calling
 	  function that the request should be denied, for informal-only notifications or
 	  acceptable, true must be returned by all the registered notification handlers.
+
+config MCUMGR_MGMT_HANDLER_USER_DATA
+	bool "MCUmgr mgmt handler user data support"
+	default n
```
Comment (nordicjm) [inline subsys/mgmt/mcumgr/mgmt/Kconfig:31]: `default n` is not needed as that is the default

File: doc/releases/release-notes-3.5.rst [hunk 2]
```diff
@@ -328,6 +328,8 @@ Libraries / Subsystems
   * Added :kconfig:option:`CONFIG_MCUMGR_GRP_IMG_ALLOW_ERASE_PENDING` that allows
     to erase slots pending for next boot, that are not revert slots.
 
+  * Added ``user_data`` as a field to :c:struct:`mgmt_handler`.
```
Comment (nordicjm) [inline doc/releases/release-notes-3.5.rst:331]:
```suggestion
  * Added ``user_data`` as an optional field to :c:struct:`mgmt_handler` when :kconfig:option:`CONFIG_MCUMGR_MGMT_HANDLER_USER_DATA` is enabled.
```

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: subsys/mgmt/mcumgr/mgmt/Kconfig (hunk 1 above)
Comment (kunoh) [reply subsys/mgmt/mcumgr/mgmt/Kconfig]: Fixed

File: doc/releases/release-notes-3.5.rst (hunk 2 above)
Comment (kunoh) [reply doc/releases/release-notes-3.5.rst]: Fixed

**Review by nordicjm — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]

**Pull request merged**